### PR TITLE
[BEAM-9291] Upload graph option in dataflow's python sdk

### PR DESCRIPTION
**Pull request by stankiewicz**
Added support for upload_graph option in dataflow's python sdk. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 1]
```diff
@@ -562,6 +562,12 @@ def _add_argparse_args(cls, parser):
         default=None,
         choices=['COST_OPTIMIZED', 'SPEED_OPTIMIZED'],
         help='Set the Flexible Resource Scheduling mode')
+    parser.add_argument(
```
Comment (aaltay) [inline sdks/python/apache_beam/options/pipeline_options.py]: Looking at the Java implementation (https://github.com/apache/beam/pull/7047), this is an experiment and not a top level option.

I believe you can handle this similar to other experiments in internal/apiclient.py

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 2]
```diff
@@ -539,6 +541,15 @@ def run_pipeline(self, pipeline, options):
     # Get a Dataflow API client and set its options
     self.dataflow_client = apiclient.DataflowApplicationClient(options)
 
+    if self.job.options.view_as(GoogleCloudOptions).upload_graph:
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Similarly, this staging can happen in apiclient.py if the experiment is present.

**Review by aaltay — CHANGES_REQUESTED** [review]

**Comment by aaltay** [comment]
I added a few comments. Could we also verify that this works as expected on Dataflow?

File: sdks/python/apache_beam/options/pipeline_options.py (hunk 1 above)
Comment (stankiewicz) [reply sdks/python/apache_beam/options/pipeline_options.py]: Re option, ok, I will handle it as experiment.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 2 above)
Comment (stankiewicz) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Re hiding it in apiclient - in first commit I've implemented it that way then I've noticed java implementation where it is implemented  in runner instead of apiclient and decided to move.  Will move again and rewrite tests.

**Comment by stankiewicz** [comment]
I have one challenge with full e2e test - python version stages full apache-beam package.
pip download 2.20dev0 fails. 
I can try to patch locally 2.19 tag just with this PR or maybe there is other way to test runner?
When I change versions to 2.19 then eventually I have error some errors on version clash (probably haven't changed everywhere), but I can see that pipeline.json is uploaded, it contains steps and graph is rendered properly in console.

**Comment by stankiewicz** [comment]
@aaltay 
Before: "The job graph is too large. Please try again with a smaller job graph, or split your job into two or more smaller jobs.", 400,  during REST submit
After: "Workflow failed. Causes: The job graph is too large. Please try again with a smaller graph." on runtime..

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 3]
```diff
@@ -563,6 +563,7 @@ def _add_argparse_args(cls, parser):
         choices=['COST_OPTIMIZED', 'SPEED_OPTIMIZED'],
         help='Set the Flexible Resource Scheduling mode')
 
+
```
Comment (aaltay) [inline sdks/python/apache_beam/options/pipeline_options.py]: Could you revert the change to this file?

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 4]
```diff
@@ -539,6 +539,7 @@ def run_pipeline(self, pipeline, options):
     # Get a Dataflow API client and set its options
     self.dataflow_client = apiclient.DataflowApplicationClient(options)
 
+
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Could you revert the change to this file?

**Comment by stankiewicz** [comment]
removed empty lines.

**Review by aaltay — APPROVED** [review]

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
Not sure if tests will start or not. This has been an ongoing problem for a while.

**Comment by stankiewicz** [comment]
retest this please

**Comment by stankiewicz** [comment]
retest this please

**Comment by stankiewicz** [comment]
I've squashed commits.

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by stankiewicz** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by stankiewicz** [comment]
retest this please

**Comment by aaltay** [comment]
Run Python PreCommit

**Comment by aaltay** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by stankiewicz** [comment]
Run Python PreCommit

**Comment by stankiewicz** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by stankiewicz** [comment]
Precommit fails due to change from #10826 
Fix will be applied in #10856 , after that retest is needed.

**Comment by stankiewicz** [comment]
Run Python PreCommit

**Comment by rohdesamuel** [comment]
> Precommit fails due to change from #10826
> Fix will be applied in #10856 , after that retest is needed.

Sorry for the breakage!

**Comment by stankiewicz** [comment]
No worries:) from what @aaltay written, there are more reasons for failuring tests

**Comment by stankiewicz** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
run python precommit

**Comment by stankiewicz** [comment]
@aaltay , can we merge?
